homeassistant/components/sun.py

```diff
@@ -51,172 +51,143 @@ def is_on(hass, entity_id=None):
     return hass.states.is_state(entity_id, STATE_ABOVE_HORIZON)
 
 
-def next_dawn(hass, entity_id=None):
+def next_dawn(hass, entity_id=None, offset=None):
     """Local datetime object of the next dawn.
 
     Async friendly.
     """
-    utc_next = next_dawn_utc(hass, entity_id)
+    utc_next = next_dawn_utc(hass, entity_id, offset)
 
     return dt_util.as_local(utc_next) if utc_next else None
 
 
-def next_dawn_utc(hass, entity_id=None):
+def next_dawn_utc(hass, entity_id=None, offset=None):
     """UTC datetime object of the next dawn.
 
     Async friendly.
     """
-    entity_id = entity_id or ENTITY_ID
-
-    state = hass.states.get(ENTITY_ID)
-
-    try:
-        return dt_util.parse_datetime(
-            state.attributes[STATE_ATTR_NEXT_DAWN])
-    except (AttributeError, KeyError):
-        # AttributeError if state is None
-        # KeyError if STATE_ATTR_NEXT_DAWN does not exist
+    entity = hass.data.get(entity_id or ENTITY_ID)
+    if entity is None:
         return None
 
+    return Sun.get_next_solar_event(
```

Comment:
Here's one example of how the current logic is failing: The user has configured an automation to happen 2 hours before sunset. The user starts up hass one hour before the next sunset. The automation helper looks at the time of the next sunset (1 hour from now), and adds 24 hours, so the event will trigger 23 hours in the future. In 23 hours, the event triggers, and hass goes to look up the next event time. If sunsets are getting later in the day, the next sunset will be in 2 hours + 3 minutes, so hass schedule the next trigger for 3 minutes later. The problem here is that adding 24 hours to a previous sunset doesn't give the time of the next sunset, although it's been close enough for us to skirt by for a while.

This is the same example with the correct logic: The user starts up hass one hour before the next sunset. The automation helper looks at the time of the next sunset (1 hour from now), and since the next sunset + offset has already occurred, it looks up the next sunset, which is in 25 hours + 3 minutes, and adds the offset, so the event will trigger 23 hours + 3 minutes in the future.